flow/raster_cache.cc

```diff
@@ -175,25 +176,57 @@ std::unique_ptr<RasterCacheResult> RasterCache::RasterizeDisplayList(
 
 void RasterCache::Prepare(PrerollContext* context,
                           Layer* layer,
-                          const SkMatrix& ctm) {
-  RasterCacheKey cache_key(layer->unique_id(), RasterCacheKeyType::kLayer, ctm);
-  Entry& entry = cache_[cache_key];
+                          const SkMatrix& ctm,
+                          RasterCacheLayerStrategy strategy) {
+  auto cache_key_optional =
+      TryToMakeRasterCacheKeyForLayer(layer, strategy, ctm);
+  if (!cache_key_optional) {
+    return;
+  }
+  Entry& entry = cache_[cache_key_optional.value()];
   entry.access_count++;
   entry.used_this_frame = true;
   if (!entry.image) {
-    entry.image = RasterizeLayer(context, layer, ctm, checkerboard_images_);
+    entry.image =
+        RasterizeLayer(context, layer, strategy, ctm, checkerboard_images_);
+  }
+}
+
+std::optional<RasterCacheKey> RasterCache::TryToMakeRasterCacheKeyForLayer(
+    const Layer* layer,
+    RasterCacheLayerStrategy strategy,
+    const SkMatrix& ctm) const {
+  switch (strategy) {
+    case RasterCacheLayerStrategy::kLayer:
+      return RasterCacheKey(layer->unique_id(), RasterCacheKeyType::kLayer,
+                            ctm);
+    case RasterCacheLayerStrategy::kLayerChildren:
+      FML_DCHECK(layer->as_container_layer());
+      auto& children_layers = layer->as_container_layer()->layers();
+      auto children_count = children_layers.size();
+      if (children_count == 0) {
+        return std::nullopt;
+      }
+      std::vector<uint64_t> ids;
+      std::transform(children_layers.begin(), children_layers.end(),
+                     std::back_inserter(ids), [](auto& layer) -> uint64_t {
+                       return layer->unique_id();
```

Comment:
I believe this is incorrect for display list layers. These layers get a new unique ID on each frame, even if the underlying display list has the same unique ID. This causes us to thrash the raster cache if you have a PictureLayer (display list) as child of something like an opacity that uses this strategy.

Reply:
Before refactoring, if OpacityLayer cached its own children, it was obtained through GetCacheableChildLayer. If the OpacityLayer contains a DisplaylistLayer, then finally it also uses the unique_id of the DisplayListLayer as the key. So the original implementation was not correct, right?

c.f.
https://github.com/flutter/engine/blob/88b7f7252c7cad276d84fa01b2af186165f8e49a/flow/layers/container_layer.cc#L252

c.f.
https://github.com/flutter/engine/blob/88b7f7252c7cad276d84fa01b2af186165f8e49a/flow/layers/container_layer.cc#L252

Reply:
Yeah, it didn't work before as well. This is unfortunate but not a regression. I'll file a new bug to track